### ISSUE TITLE
Make runs only one instance of Cumulative targets

### ISSUE THREAD

**Issue by theory** [Homebrew/brew#4848]
Summary
---------

Homebrew's `make` command appears to execute only a single instance of a Makefile target when it is specified multiple times with double colons. This appears to happen only when run from a formula, never when run manually on the command-line. I've created a [test tap](/theory/homebrew-makefail) to try to simplify the recreation of the issue.

Survey
-------

- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [x] ran a `brew` command and reproduced the problem with multiple formulae? [I've tested both with a hand-edited formula and with the [test tap](/theory/homebrew-makefail)]
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew config` and `brew doctor` and included their output with your issue?

Configuration
-------------

```
$ brew config
HOMEBREW_VERSION: 1.7.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: 61ace27006a1cb444552c0be10087fcbc3051643
Last commit: No signature
3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: f6e820d7dac23c14aed1be2067986c081f41b83a
Core tap last commit: 75 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_GIT: git
CPU: quad-core 64-bit broadwell
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1000
Git: 2.17.0 => /usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: N/A
macOS: 10.13.6-x86_64
CLT: 10.0.0.0.1.1527767617
CLT headers: N/A
Xcode: N/A
XQuartz: N/A

$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libcrypto.1.1.dylib
  /usr/local/lib/libgmp.10.dylib
  /usr/local/lib/libhogweed.4.4.dylib
  /usr/local/lib/libnettle.6.4.dylib
  /usr/local/lib/libssl.1.1.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/gmp.h
  /usr/local/include/nettle/aes.h
  /usr/local/include/nettle/arcfour.h
  /usr/local/include/nettle/arctwo.h
  /usr/local/include/nettle/asn1.h
  /usr/local/include/nettle/base16.h
  /usr/local/include/nettle/base64.h
  /usr/local/include/nettle/bignum.h
  /usr/local/include/nettle/blowfish.h
  /usr/local/include/nettle/buffer.h
  /usr/local/include/nettle/camellia.h
  /usr/local/include/nettle/cast128.h
  /usr/local/include/nettle/cbc.h
  /usr/local/include/nettle/ccm.h
  /usr/local/include/nettle/cfb.h
  /usr/local/include/nettle/chacha-poly1305.h
  /usr/local/include/nettle/chacha.h
  /usr/local/include/nettle/ctr.h
  /usr/local/include/nettle/curve25519.h
  /usr/local/include/nettle/des-compat.h
  /usr/local/include/nettle/des.h
  /usr/local/include/nettle/dsa-compat.h
  /usr/local/include/nettle/dsa.h
  /usr/local/include/nettle/eax.h
  /usr/local/include/nettle/ecc-curve.h
  /usr/local/include/nettle/ecc.h
  /usr/local/include/nettle/ecdsa.h
  /usr/local/include/nettle/eddsa.h
  /usr/local/include/nettle/gcm.h
  /usr/local/include/nettle/gosthash94.h
  /usr/local/include/nettle/hkdf.h
  /usr/local/include/nettle/hmac.h
  /usr/local/include/nettle/knuth-lfib.h
  /usr/local/include/nettle/macros.h
  /usr/local/include/nettle/md2.h
  /usr/local/include/nettle/md4.h
  /usr/local/include/nettle/md5-compat.h
  /usr/local/include/nettle/md5.h
  /usr/local/include/nettle/memops.h
  /usr/local/include/nettle/memxor.h
  /usr/local/include/nettle/nettle-meta.h
  /usr/local/include/nettle/nettle-stdint.h
  /usr/local/include/nettle/nettle-types.h
  /usr/local/include/nettle/pbkdf2.h
  /usr/local/include/nettle/pgp.h
  /usr/local/include/nettle/pkcs1.h
  /usr/local/include/nettle/poly1305.h
  /usr/local/include/nettle/pss-mgf1.h
  /usr/local/include/nettle/pss.h
  /usr/local/include/nettle/realloc.h
  /usr/local/include/nettle/ripemd160.h
  /usr/local/include/nettle/rsa.h
  /usr/local/include/nettle/salsa20.h
  /usr/local/include/nettle/serpent.h
  /usr/local/include/nettle/sexp.h
  /usr/local/include/nettle/sha.h
  /usr/local/include/nettle/sha1.h
  /usr/local/include/nettle/sha2.h
  /usr/local/include/nettle/sha3.h
  /usr/local/include/nettle/twofish.h
  /usr/local/include/nettle/umac.h
  /usr/local/include/nettle/version.h
  /usr/local/include/nettle/yarrow.h
  /usr/local/include/openssl/aes.h
  /usr/local/include/openssl/asn1.h
  /usr/local/include/openssl/asn1_mac.h
  /usr/local/include/openssl/asn1t.h
  /usr/local/include/openssl/async.h
  /usr/local/include/openssl/bio.h
  /usr/local/include/openssl/blowfish.h
  /usr/local/include/openssl/bn.h
  /usr/local/include/openssl/buffer.h
  /usr/local/include/openssl/camellia.h
  /usr/local/include/openssl/cast.h
  /usr/local/include/openssl/cmac.h
  /usr/local/include/openssl/cms.h
  /usr/local/include/openssl/comp.h
  /usr/local/include/openssl/conf.h
  /usr/local/include/openssl/conf_api.h
  /usr/local/include/openssl/crypto.h
  /usr/local/include/openssl/ct.h
  /usr/local/include/openssl/des.h
  /usr/local/include/openssl/dh.h
  /usr/local/include/openssl/dsa.h
  /usr/local/include/openssl/dtls1.h
  /usr/local/include/openssl/e_os2.h
  /usr/local/include/openssl/ebcdic.h
  /usr/local/include/openssl/ec.h
  /usr/local/include/openssl/ecdh.h
  /usr/local/include/openssl/ecdsa.h
  /usr/local/include/openssl/engine.h
  /usr/local/include/openssl/err.h
  /usr/local/include/openssl/evp.h
  /usr/local/include/openssl/hmac.h
  /usr/local/include/openssl/idea.h
  /usr/local/include/openssl/kdf.h
  /usr/local/include/openssl/lhash.h
  /usr/local/include/openssl/md2.h
  /usr/local/include/openssl/md4.h
  /usr/local/include/openssl/md5.h
  /usr/local/include/openssl/mdc2.h
  /usr/local/include/openssl/modes.h
  /usr/local/include/openssl/obj_mac.h
  /usr/local/include/openssl/objects.h
  /usr/local/include/openssl/ocsp.h
  /usr/local/include/openssl/opensslconf.h
  /usr/local/include/openssl/opensslv.h
  /usr/local/include/openssl/ossl_typ.h
  /usr/local/include/openssl/pem.h
  /usr/local/include/openssl/pem2.h
  /usr/local/include/openssl/pkcs12.h
  /usr/local/include/openssl/pkcs7.h
  /usr/local/include/openssl/rand.h
  /usr/local/include/openssl/rc2.h
  /usr/local/include/openssl/rc4.h
  /usr/local/include/openssl/rc5.h
  /usr/local/include/openssl/ripemd.h
  /usr/local/include/openssl/rsa.h
  /usr/local/include/openssl/safestack.h
  /usr/local/include/openssl/seed.h
  /usr/local/include/openssl/sha.h
  /usr/local/include/openssl/srp.h
  /usr/local/include/openssl/srtp.h
  /usr/local/include/openssl/ssl.h
  /usr/local/include/openssl/ssl2.h
  /usr/local/include/openssl/ssl3.h
  /usr/local/include/openssl/stack.h
  /usr/local/include/openssl/symhacks.h
  /usr/local/include/openssl/tls1.h
  /usr/local/include/openssl/ts.h
  /usr/local/include/openssl/txt_db.h
  /usr/local/include/openssl/ui.h
  /usr/local/include/openssl/whrlpool.h
  /usr/local/include/openssl/x509.h
  /usr/local/include/openssl/x509_vfy.h
  /usr/local/include/openssl/x509v3.h

Warning: Unbrewed .la files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected .la files:
  /usr/local/lib/libgmp.la

Warning: Unbrewed .pc files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected .pc files:
  /usr/local/lib/pkgconfig/hogweed.pc
  /usr/local/lib/pkgconfig/libcrypto.pc
  /usr/local/lib/pkgconfig/libssl.pc
  /usr/local/lib/pkgconfig/nettle.pc
  /usr/local/lib/pkgconfig/openssl.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libcrypto.a
  /usr/local/lib/libgmp.a
  /usr/local/lib/libhogweed.a
  /usr/local/lib/libnettle.a
```

Pretty sure those extra files have no impact on this issue.

Goals
------

### To help us debug your issue please explain:

- What you were trying to do (and why)

    > I'm working on a formula that installs Perl CPAN modules via [cpanminus](https://formulae.brew.sh/formula/cpanminus). I'm doing this to create a tidy, self-contained install of [Sqitch](https://sqitch.org).

- What happened (include command output)

    > Homebrew's `make` command appears to execute only a single instance of a Makefile target when it is specified multiple times with double colons. This appears to happen only when run from a formula, never when run manually on the command-line.

- What you expected to happen

    > I expected all instances of a cumulative target specified with two colons (`::`) to execute.

### Step-by-step reproduction instructions (by running `brew` commands)

1.  Install the tap and install `makefail`:

    ``` sh
    brew tap theory/makefail
    brew install --keep-tmp makefail
    ```

    Sample output:

    ```
    ==> Installing makefail from theory/makefail
    ==> Downloading https://raw.githubusercontent.com/theory/homebrew-makefail/master/tryme.tgz
    Already downloaded: /Users/david/Library/Caches/Homebrew/downloads/36509c8de1f57856d162c29aa3219c6c50900bc352d5485bbb150fab404d727d--tryme.tgz
    ==> cpanm --quiet --notest --local-lib-contained tryit List::MoreUtils::XS Exporter::Tiny
    ==> cpanm --verbose --notest --local-lib-contained tryit --installdeps .
    Last 15 lines from /Users/david/Library/Logs/Homebrew/makefail/02.cpanm:
    Checking if you have List::MoreUtils::XS 0.426 ... Yes (0.428)
    Checking if you have ExtUtils::MakeMaker 0 ... Yes (6.66)
    Checking if you have Exporter::Tiny 0.038 ... Yes (1.002001)
    OK
    Building List-MoreUtils-0.428 ... cp lib/List/MoreUtils/Contributing.pod blib/lib/List/MoreUtils/Contributing.pod
    cp lib/List/MoreUtils.pm blib/lib/List/MoreUtils.pm
    cp lib/List/MoreUtils/PP.pm blib/lib/List/MoreUtils/PP.pm
    Appending installation info to /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/darwin-thread-multi-2level/perllocal.pod
    OK
    Successfully installed List-MoreUtils-0.428
    Installing /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-0.428/MYMETA.json
    Installing /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-0.428/install.json
    ! Installing the dependencies failed: Module 'List::MoreUtils' is not installed
    ! Bailing out the installation for ..
    1 distribution installed
    ==> Kept temporary files
    Temporary files retained at /private/tmp/makefail-20180906-25247-4rfpkj

    If reporting this issue please do so at (not Homebrew/brew or Homebrew/core):
    https://github.com/theory/homebrew-makefail/issues
    ```

2.  Note that cpanminus reports that List::MoreUtils was "successfully
    installed", but then later says that it was not. Examine the full log output
    in `~/Library/Logs/Homebrew/makefail/02.cpanm` ([sample](/theory/homebrew-makefail/blob/master/eg/fail.log)).

3.  Switch to the temp directory into which files have been installed, indicated
    by the log line that begins with "Entering". In the [sample](/theory/homebrew-makefail/blob/master/eg/fail.log),
    its `/private/tmp/makefail-20180906-25247-4rfpkj/tryme`. Note  that
    `MoreUtils.pm` has not been copied into the `tryit` directory.

    ```
    $ find tryit -type f
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-XS-0.428/MYMETA.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-XS-0.428/install.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/XSLoader-0.24/MYMETA.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/XSLoader-0.24/install.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/Exporter-Tiny-1.002001/MYMETA.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/Exporter-Tiny-1.002001/install.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-0.428/MYMETA.json
    tryit/lib/perl5/darwin-thread-multi-2level/.meta/List-MoreUtils-0.428/install.json
    tryit/lib/perl5/darwin-thread-multi-2level/auto/List/MoreUtils/XS/XS.bundle
    tryit/lib/perl5/darwin-thread-multi-2level/auto/List/MoreUtils/XS/XS.bs
    tryit/lib/perl5/darwin-thread-multi-2level/auto/List/MoreUtils/XS/.packlist
    tryit/lib/perl5/darwin-thread-multi-2level/auto/Exporter/Tiny/.packlist
    tryit/lib/perl5/darwin-thread-multi-2level/auto/XSLoader/.packlist
    tryit/lib/perl5/darwin-thread-multi-2level/List/MoreUtils/XS.pm
    tryit/lib/perl5/darwin-thread-multi-2level/perllocal.pod
    tryit/lib/perl5/darwin-thread-multi-2level/XSLoader.pm
    tryit/lib/perl5/Exporter/Tiny.pm
    tryit/lib/perl5/Exporter/Tiny/Manual/Exporting.pod
    tryit/lib/perl5/Exporter/Tiny/Manual/QuickStart.pod
    tryit/lib/perl5/Exporter/Tiny/Manual/Etc.pod
    tryit/lib/perl5/Exporter/Tiny/Manual/Importing.pod
    tryit/lib/perl5/Exporter/Shiny.pm
    ```

4.  Switch to the work directory reported in `02.cpanm` and then into the
    `List-MoreUtils-0.428` directory and examine `List-MoreUtils-0.428/Makefile`
    ([sample](/theory/homebrew-makefail/blob/master/eg/Makefile)). Note that there are two `pure_site_install`
    targets, both specified with a double colon, so `make` should know to run
    them both. They look like this:

    ``` make
    pure_site_install ::
    	$(NOECHO) $(RM_F) '$(DESTINSTALLSITEARCH)/List/MoreUtils.pm'
    ```

    And:

    ```
    pure_site_install :: all
    	$(NOECHO) $(MOD_INSTALL) \
    		read $(SITEARCHEXP)/auto/$(FULLEXT)/.packlist \
    		write $(DESTINSTALLSITEARCH)/auto/$(FULLEXT)/.packlist \
    		$(INST_LIB) $(DESTINSTALLSITELIB) \
    		$(INST_ARCHLIB) $(DESTINSTALLSITEARCH) \
    		$(INST_BIN) $(DESTINSTALLSITEBIN) \
    		$(INST_SCRIPT) $(DESTINSTALLSITESCRIPT) \
    		$(INST_MAN1DIR) $(DESTINSTALLSITEMAN1DIR) \
    		$(INST_MAN3DIR) $(DESTINSTALLSITEMAN3DIR)
    	$(NOECHO) $(WARN_IF_OLD_PACKLIST) \
    		$(PERL_ARCHLIB)/auto/$(FULLEXT)
    ```

    The second one is the one that should install the `MoreUtils.pm`. It does
    not appear to have been executed.

5.  Run `make install` manually:

    ```
    /usr/local/Homebrew/Library/Homebrew/shims/mac/super/make install
    Installing /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/List/MoreUtils.pm
    Installing /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/List/MoreUtils/PP.pm
    Installing /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/List/MoreUtils/Contributing.pod
    Appending installation info to /private/tmp/makefail-20180906-25247-4rfpkj/tryme/tryit/lib/perl5/darwin-thread-multi-2level/perllocal.pod
    ```

6.  Note that `MoreUtils.pm` has not been installed. Questions:

    *   Why didn't it get installed by the execution of the formula?
    *   Is there something in the environment that changes the behavior of `make`?
    *   Is it the case that `make` only runs one `pure_site_install` target when run by the formula?

7.  Try again, but turn on debugging output for `make install`:

    ```
    brew install --keep-tmp --with-install-debug makefail
    ```

    The build appears to succeed, but only because of the way `cpanminus` has
    been executed. `MoreUtils.pm` still will not have been installed. See
    `~/Library/Logs/Homebrew/makefail/02.cpanm` for the complete debug output
    [sample](/theory/homebrew-makefail/blob/master/eg/debug.log).

I don't have the make-foo knowledge to understand the debugging output. :-(

**Comment by theory** [comment]
I have since added `--with-install-verbose`, which turns on the Perl installation verbosity. It looks like both `pure_site_install` targets are called after all, it's just that the modules aren't installed at all. So now I'm less inclined to think it's an issue with Homebrew's `make`, though it's still weird that I can't replicate the issue when executing the various steps manually. That might be down to something in cpanminus, too. Still futzing with it, and apologies for the noise as I try to work it out.

**Comment by theory** [comment]
Finally replicated outside of Homebrew, after dumping Homebrew's environment and adding it to a script. Then I eliminated variables until I found the one that was the problem. This one-liner exhibits the same bug:

``` sh
env MAKEFLAGS=-j4 cpanm --verbose --notest -L tryit --install-args 'NOECHO= VERBOSE=1' List::MoreUtils
```

So I don't think this is a Homebrew bug now. So far I can only replicate it with the macOS system Perl, and not any custom Perl builds, so that's weird, but nothing to do with Homebrew. Appreciate your forbearance as I worked through this issue.
